src/lightning/pytorch/trainer/connectors/data_connector.py

```diff
@@ -251,8 +251,11 @@ def _get_distributed_sampler(
     """This function is used to created the distributed sampler injected within the user DataLoader."""
     kwargs["shuffle"] = shuffle and not overfit_batches
     kwargs.setdefault("seed", int(os.getenv("PL_GLOBAL_SEED", 0)))
-    cls = UnrepeatedDistributedSamplerWrapper if mode == RunningStage.PREDICTING else DistributedSamplerWrapper
-    return cls(dataloader.sampler, **kwargs)
+    if mode == RunningStage.PREDICTING:
+        return UnrepeatedDistributedSamplerWrapper(dataloader.sampler, **kwargs)
+    if isinstance(dataloader.sampler, (RandomSampler, SequentialSampler)):
```

Comment:
I would suggest that this `if` is put in the `DistributedSamplerWrapper` __init__.

It's docstring advertises it as

> Allows you to use any sampler in distributed mode. It will be automatically used by Lightning in distributed mode if sampler replacement is enabled.

Putting it here invalidates that statement.

Reply:
The `DistributedSamplerWrapper` is a confusing class. Putting this condition into its init won't be enough. Its implementation would have to be changed. It is much more sane to put this logic here, because the PyTorch Distributed sampler supports handling these two.

Reply:
My point is that that class is meant to encapsulate this kind of logic. The `DistributedSamplerWrapper` is itself a `DistributedSampler`. If one was using it standalone (possible, as it's public API), then the reported issue should be fixed for it too.

I think the ideal fix would look into why it has this problem, but the if-else workaround proposed here is good to quickly fix the issue. The implementation changes would be minimal as you could check for `isinstance(_DatasetSamplerWrapper)` to split the logic

Reply:
> I think the ideal fix would look into why it has this problem

It is obvious why this is the case. The custom-distributed sampler is bypassing the entire seeding logic in the original implementation, which is not a problem if you set a global seed. I disagree with your proposal, there is no point in using the custom logic for the standard samplers in PyTorch. Can you open a PR yourself to redesign the custom class to your liking?

Reply:
Approved to unblock